File: codebox/面试测试/Sudoku_solver.cpp
bool checkValid(vector<string>& board, int x, int y){
	bool flags[9] = {0};
	//列有效
	for(int i = 0; i < 9; ++i){
		if(board[x][i] >= '1' && board[x][i] <= '9'){
			if(!flags[board[x][i] - '1'])
				flags[board[x][i] - '1'] = true;
			else
				return false;
		}
	}
	//行有效
	memset(flags, 0, 9);
	for(int i = 0; i < 9; ++i){
		if(board[i][y] >= '1' && board[i][y] <= '9'){
			if(!flags[board[i][y] - '1'])
				flags[board[i][y] - '1'] = true;
			else
				return false;
		}
	}
	//小九宫格有效
	int xx = x/3*3;
	int yy = y/3*3;
	memset(flags, 0, 9);
	for(int i = 0; i < 3; ++i){
		for(int j = 0; j < 3; ++j){
			if(board[xx+i][yy+j] >= '1' && board[xx+i][yy+j] <= '9'){
				if(!flags[board[xx+i][yy+j]-'1'])
					flags[board[xx+i][yy+j]-'1'] = true;
				else
					return false;
			}
		}
	}
	return true;
}
bool solve(vector<string>&board){
	for(int i = 0; i < 9; i++){
		for(int j = 0; j < 9; j++){
			if(board[i][j] != '.')
				continue;
			for(char k = '1'; k <= '9'; k++){//填入一个数
				board[i][j] = k;
				if(checkValid(board,i,j)){//有效则继续填
					if(solve(board))
						return true;
				}
				board[i][j] = '.';//无效则还原
			}
			return false;
		}
	}
	return true;
}

File: codebox/leetcode/Sudoku_Solver.java
package main;

import java.util.HashSet;
import java.util.Set;

public class Sudoku_Solver {
  
    public static boolean isValidSudoku(char value,int row,int col,char[][] board){//specific
        boolean res = true;
        for(int i=0;i<9;i++){
            //check row
            if(board[row][i]==value){
                res = false;
                break;
            }
            //check col
            if(board[i][col]==value){
                res = false;
                break;
            }
            //check square
            int row_s = 3*(int)(row/3) + (int)(i/3);
            int col_s = 3*(int)(col/3) + i%3;
            if(board[row_s][col_s]==value){
                res = false;
                break;
            }
        }
        return res;       
    }
    
    public static boolean solve(char[][] board){
	boolean isfind=false;
	int i=0;int j=0;
	for(i=0;i<board.length;i++){
	    for(j=0;j<board[i].length;j++){
		if(board[i][j]=='.'){
		    isfind=true;
		    break;
		}
	    }
	    if(isfind)
		break;
	}
	if (isfind) {
	    for (int v = 1; v <= 9; v++) {
		char value=(char) ('0' + v);
		if (isValidSudoku(value,i,j,board)) {
		    board[i][j] = value;
		    if(solve(board))
			return true;
		}
	    }
	    board[i][j]='.';
	    return false;
	}
	else
	    return true;
    }
    public static void solveSudoku(char[][] board) {
        // Start typing your Java solution below
        // DO NOT write main() function
	solve(board);
	
    }
    public static void main(String args[]){
	String tt[]={"..9748...","7........",".2.1.9...","..7...24.",".64.1.59.",".98...3..","...8.3.2.","........6","...2759.."};
	char num[][]=new char[tt.length][tt[0].length()];
	for(int i=0;i<tt.length;i++){
	    for(int j=0;j<tt[0].length();j++){
		num[i][j]=tt[i].charAt(j);
	    }
	}
	solveSudoku(num);
	System.out.println("11");
	
    }

}    


